### ISSUE TITLE
Add string functions similar to BASIC

### ISSUE THREAD

**Issue by Thraka** [neilsf/XC-BASIC#115]
String functions should be added that implement a lot of the standard BASIC string functions without having to do workarounds with array buffers.

**Comment by neilsf** [comment]
Hi. It's true that string functions are a bit bizarre in XC=BASIC, they rather follow the C way of handling strings. The problem is that two major features are required for implementing string functions similar to CBM BASIC:

1. dynamic memory allocation (to be able to create a string in runtime)
2. garbage collection (to free up memory when a string is no longer needed)

Both would make the program extremely slow, just like they make CBM BASIC very very slow. In CBM BASIC you may run into a situation where garbage collection runs for minutes. XC=BASIC is meant to be fast, so for this reason dynamic strings had to be sacrificed.

If there's a third way you can suggest, i.e. both fast and easy, I'm open to implementing a different approach in the future.

**Comment by Thraka** [comment]
What if you had a 256 buffer allocated behind the scenes as string working space? Then the functions that do string manipulation do it in this other buffer and return the results the target string buffer?

This still requires that a user think about their strings and allocate the potential max for their string, but then gives them the flexibility to not totally manage it with multiple lines of code.

**Comment by tomxp411** [comment]
It seems to me that the best choice is to pre-define strings with a length (ie: a character array.) Then when doing any sort of string operations (A$ + B$, LEFT$(A$,3), etc), the result is placed in the already-allocated variable. A behind the scenes buffer might be helpful, but I'd probably build it as a stack, rather than a single buffer. An expression like A$=RIGHT$("00000"+C$,5) may end up needing more than one buffer, depending on how it's parsed.

So my suggestion is to only allow LValues (the target of an assignment) to be pre-declared, fixed length string values, and build a stack using the combined length of both strings when using the concatenation operator. (I recommend &, as it's what VB uses.) 

If you use a stack as part of the expression parsing, then you can clean it up by simply moving the stack pointer back up to the top when the expression is fully parsed out. And since string variables would be pre-declared and static, there would be no need for garbage collection.

**Comment by neilsf** [comment]
@Thraka @tomxp411 Indeed, a stack is needed if the expression contains multiple operations and the result of one operation is the operand of another. However I'm still concerned about speed. A stack involves pushing and pulling values which is very fast when it's about 2-byte integers but copying (max) 256 bytes there and back will be slow unless it's not optimized in some way. 

I'll definitely think about it cause most of the people who tried XC=BASIC say that string handling is very difficult and they may be right. I may redesign this part but I won't start tonight as my head is falling onto the desk now :-) Thanks for your input and let's keep this thread alive. I would be interested in how other languages do it, especially the ones targeted to 8-bit machines. Is it worth checking in Pascal for example?

**Comment by EgonOlsen71** [comment]
> In CBM BASIC you may run into a situation where garbage collection runs for minutes.

But that's only the case because it's not very well implemented. It's O(n*n) (with n being the number of string variables). But it doesn't has to be that way. If you, for example, add back references to the strings in memory, the run time is down to O(n) with a proper implementation.
